pkg/rpc/context.go

```diff
@@ -2719,26 +2748,224 @@ func (rpcCtx *Context) loadOrCreateConnAttempt(
 	// happen if our previous connect attempt failed between pings. Without this
 	// protection we would continually try opening new dialback connections, but
 	// never observe the result.
-	if previousAttempt := rpcCtx.dialbackMu.m[nodeID]; previousAttempt != nil {
-		select {
-		case <-previousAttempt.initialHeartbeatDone:
-			// The connection attempt was completed, return the outcome of it.
-			err, _ := previousAttempt.err.Load().(error)
-			if err == nil {
-				// If it completed without error then don't track the connection
-				// anymore. If it did have an error we need to track it until it later gets cleared.
-				rpcCtx.dialbackMu.m[nodeID] = nil
+	err := previousAttempt.Health()
+	if err == nil {
+		// If it completed without error then don't track the connection anymore and
+		// return success.
+		rpcCtx.dialbackMu.m[nodeID] = nil
+		return nil
+	}
+	// If still don't know the outcome of the previous attempt, allow this attempt
+	// to continue and check again in the future.
+	if errors.Is(err, ErrNotHeartbeated) {
+		return nil
+	}
+
+	return err
+}
+
+// hasSiblingConn takes a peer as identified by `self` and tries to find a sibling:
+// - if self.NodeID != 0, finds another conn with matching NodeID but different TargetAddr.
+// - if self.NodeID == 0, finds another conn with matching TargetAddr.
+//
+// In both cases, if such a conn exists that became healthy *after* ours became
+// unhealthy, `healthy` will be true. If no such conn exists, (false, false) is
+// returned.
+//
+// TODO(during review): move after dust has settled.
+func hasSiblingConn(peers map[peerKey]*peer, self peerKey) (healthy, ok bool) {
```

Comment:
Needs to be named *Locked to show that we must lock peers prior to calling this as `shouldDeleteAfter` does.